src/IndexMaps/realindexmap.jl

```diff
@@ -75,12 +76,113 @@ function calculate_ind_values(imap::RealIndexMap, xs::Vector, dims::Vector{Int})
+    throw(
+      "expected a two-element vector [a,b] with 0≤a<b≤1 as input span, instead found $span"
+    )
+  end
+
+  dims = dim.(dimension_inds(imap, d))
+  @assert all(y -> y == first(dims), dims)
+  base = first(dims)
+  L = length(dimension_inds(imap, d))
+
+  if !exact_grid
+    grid_points = range(span[1], span[2] - (span[2] - span[1]) / N; length=N)
+  else #exact_grid = true
+    if span[1] == 0 && span[2] == 1
+      inv_step = min(Int(floor(log(base, N))), L)
+      if !(log(base, N) ≈ inv_step)
+        @warn "rounding $N down to $(Int(base^inv_step)) exact gridpoints!"
+      end
+      # now 1/inv_step should provide the step
+      grid_points = collect(span[1]:(1 / base^inv_step):span[2])[1:(end - 1)]
+
+    else
+
+      #get a count of the number of gridpoints within the span
+      points_in_span = floor(span[end] * base^L) - ceil(span[1] * base^L) + 1
+      #TODO: figure out a way to calculate this without exponentiating to the power of L
+
+      #exclude the endpoint in the count
+      if floor(span[end] * base^L) == span[end] * base^L
+        points_in_span = points_in_span - 1
+      end
+
+      if points_in_span <= 0
+        @warn "No exact gridpoints found in this span!"
+        grid_points = []
+      else
+        oldN = N
+        stepsize = ceil(points_in_span / N)
+        N = Int(floor(points_in_span / stepsize)) + 1
+        startval = ceil(span[1] * base^L) / base^L #startval is the smallest grid point ≥ span[1]
+
+        if startval + (N - 1) * stepsize / base^L >= span[2]
+          N = N - 1
+        end
+
+        if N < oldN
+          @warn "rounding $oldN down to $N exact gridpoints to maintain equal spacing."
+        end
+
+        grid_points = [startval + i * (stepsize / base^L) for i in 0:(N - 1)]
+      end
+    end
+  end
+
+  #add enforced points in naively
+  for point in enforced
+    if point >= span[2] || point < span[1]
+      @warn "enforced point $point outside of span $span, not including it"
+      continue
+    end
+    nearest_idx = searchsortedfirst(grid_points, point)
+    if nearest_idx > length(grid_points) || grid_points[nearest_idx] != point
+      insert!(grid_points, nearest_idx, point)
+    end
+  end
+
+  return grid_points
+end
+
+function grid_points(imap::RealIndexMap, d::Int; kwargs...)
+  dims = dim.(dimension_inds(imap, d))
+  @assert all(y -> y == first(dims), dims)
+  base = dims[d]
+  L = length(dimension_inds(imap, d))
+  return grid_points(imap, base^L, d; kwargs...)
+end
+
+""" 
+  Picks a random grid point from `imap` given a dimension
+"""
+function rand_p(rng::AbstractRNG, imap::RealIndexMap, d::Integer)
   dims = dim.(dimension_inds(imap, d))
   @assert all(y -> y == first(dims), dims)
-  base = float(first(dims))
+  base = dims[d]
   L = length(dimension_inds(imap, d))
-  a = round(base^L / N)
-  grid_points = [i * (a / base^L) for i in 0:(N + 1)]
-  return filter(x -> x < 1, grid_points)
+  return rand(rng, 0:(base^L - 1)) / base^L
```

Comment:
Thanks for adding these back! It seems this will fail for large `L`

What about randomly choosing `L` numbers each from `[0,base)` then adding each up (potentially using some helper function we already have, maybe `calculate_p`)?

Reply:
I found that I could get this to work for large L by doing `rand(rng, 0:(big(base)^L - 1)) / big(base)^L`